### PR TITLE
[opt] Better encapsulate BLS bounds

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #2177, #2327 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/ir/scratch_pad.h [hunk 1]
```diff
@@ -224,8 +234,6 @@ class ScratchPads {
   void print() {
     for (auto &it : pads) {
       TI_P(it.first->node_type_name);
-      TI_P(it.second.bounds[0]);
-      TI_P(it.second.bounds[1]);
```
Comment (xumingkuan) [inline taichi/ir/scratch_pad.h:228]: nit: If this function is merely used for debugging, why not printing bounds as well?

File: taichi/transforms/make_block_local.cpp [hunk 2]
```diff
@@ -67,45 +68,58 @@ void make_block_local_offload(OffloadedStmt *offload,
...
             i, j, k = bls_to_global(bls_element_id)
             bls[bls_element_id] = x[i, j, k]
             // or x[i, j, k] = bls[bls_element_id]
             bls_element_id += block_dim;
           }
 
+          func bls_to_global(bls_element_id):
+            partial = bls_element_id
+            global_indices = []  // "i, j, k"
+            for i in [(dim - 1), 0]:
```
Comment (xumingkuan) [inline taichi/transforms/make_block_local.cpp:91]: nit:
```suggestion
            for i in reversed(range(0, dim)):
```
`[(dim - 1), 0]` itself has a different meaning in Python so I think we should avoid it, and `range(dim - 1, -1, -1)` (if we replace it with this one) is a little bit hard to understand.

File: taichi/transforms/make_block_local.cpp [hunk 3]
```diff
@@ -67,45 +68,58 @@ void make_block_local_offload(OffloadedStmt *offload,
...
+              pad_size = pad.pad_size[i]  // a.k.a. bounds[i].range()
+              bls_coord = partial % pad_size
+              partial = partial / pad_size
+              global_index_at_i = BlockerCorner[i] + bls_coord
```
Comment (xumingkuan) [inline taichi/transforms/make_block_local.cpp:95]:
```suggestion
              global_index_at_i = BlockCorner[i] + bls_coord
```

**Review by xumingkuan — APPROVED** [review]
LGTM in general!

File: taichi/ir/scratch_pad.h (hunk 1 above)
Comment (k-ye) [reply taichi/ir/scratch_pad.h:228]: Fixed :)

**Pull request merged**